// ngx-tetris/esm2015/classes/Piece.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/** @enum {number} */
const PieceRotation = {
    DEG_0: 0,
    DEG_90: 1,
    DEG_180: 2,
    DEG_270: 3,
};
export { PieceRotation };
PieceRotation[PieceRotation.DEG_0] = 'DEG_0';
PieceRotation[PieceRotation.DEG_90] = 'DEG_90';
PieceRotation[PieceRotation.DEG_180] = 'DEG_180';
PieceRotation[PieceRotation.DEG_270] = 'DEG_270';
/** @enum {number} */
const PieceTypes = {
    Dot: 1,
    Box: 2,
    Line: 3,
    T: 4,
    L: 5,
    Lr: 6,
    Z: 7,
    S: 8,
};
export { PieceTypes };
PieceTypes[PieceTypes.Dot] = 'Dot';
PieceTypes[PieceTypes.Box] = 'Box';
PieceTypes[PieceTypes.Line] = 'Line';
PieceTypes[PieceTypes.T] = 'T';
PieceTypes[PieceTypes.L] = 'L';
PieceTypes[PieceTypes.Lr] = 'Lr';
PieceTypes[PieceTypes.Z] = 'Z';
PieceTypes[PieceTypes.S] = 'S';
/** @enum {string} */
const PieceColors = {
    Box: 'color-box',
    Dot: 'color-dot',
    Line: 'color-line',
    T: 'color-t',
    L: 'color-l',
    Lr: 'color-lr',
    Z: 'color-z',
    S: 'color-s',
};
export { PieceColors };
export class Piece {
    /**
     * @param {?} x
     * @param {?} y
     * @param {?} gridSize
     * @param {?} maps
     */
    constructor(x, y, gridSize, maps) {
        this.color = 'red';
        this.rotation = PieceRotation.DEG_0;
        this.map = [
            [], []
        ];
        this._lastConfig = null;
        this._gridSize = gridSize;
        this.x = x;
        this.y = y;
        this._maps = maps;
        this.map = this._maps[this.rotation];
    }
    /**
     * @return {?}
     */
    get positionsOnGrid() {
        /** @type {?} */
        const acc = [];
        for (let row = 0; row < 4; row++) {
            for (let col = 0; col < 4; col++) {
                if (this.map[row][col]) {
                    /** @type {?} */
                    const pos = (this.y + row) * this._gridSize.width + this.x + col;
                    if (pos > 0) {
                        acc.push(pos);
                    }
                }
            }
        }
        return acc;
    }
    /**
     * @return {?}
     */
    store() {
        this._lastConfig = {
            x: this.x,
            y: this.y,
            rotation: this.rotation,
            map: this.map
        };
    }
    /**
     * @return {?}
     */
    clearStore() {
        this._lastConfig = null;
    }
    /**
     * @return {?}
     */
    revert() {
        if (this._lastConfig) {
            for (const x in this._lastConfig) {
                if (this._lastConfig.hasOwnProperty(x)) {
                    this[x] = this._lastConfig[x];
                }
            }
            this._lastConfig = null;
        }
    }
    /**
     * @return {?}
     */
    rotate() {
        /** @type {?} */
        const keys = Object.keys(this._maps);
        /** @type {?} */
        let idx = keys.indexOf(this.rotation.toString());
        if (idx >= keys.length - 1) {
            this.rotation = /** @type {?} */ (keys[0]);
        }
        else {
            this.rotation = /** @type {?} */ (keys[++idx]);
        }
        this.map = this._maps[this.rotation];
    }
    /**
     * @return {?}
     */
    moveDown() {
        this.y++;
    }
    /**
     * @return {?}
     */
    moveRight() {
        this.x++;
    }
    /**
     * @return {?}
     */
    moveLeft() {
        this.x--;
    }
    /**
     * @return {?}
     */
    get bottomRow() {
        return this.y + 3;
    }
    /**
     * @return {?}
     */
    get rightCol() {
        /** @type {?} */
        let col = 3;
        while (col >= 0) {
            for (let row = 0; row <= 3; row++) {
                if (this.map[row][col]) {
                    return this.x + col;
                }
            }
            col--;
        }
    }
    /**
     * @return {?}
     */
    get leftCol() {
        return this.x;
    }
}
if (false) {
    /** @type {?} */
    Piece.prototype.color;
    /** @type {?} */
    Piece.prototype.x;
    /** @type {?} */
    Piece.prototype.y;
    /** @type {?} */
    Piece.prototype.rotation;
    /** @type {?} */
    Piece.prototype._maps;
    /** @type {?} */
    Piece.prototype.map;
    /** @type {?} */
    Piece.prototype._gridSize;
    /** @type {?} */
    Piece.prototype._lastConfig;
}
