// ngx-tetris/esm2015/services/game-manager.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { PiecesFactory } from '../classes/PiecesFactory';
import { Subject } from 'rxjs/internal/Subject';
/** @type {?} */
const SPAWN_POSITION_X = 4;
/** @type {?} */
const SPAWN_POSITION_Y = -4;
export class Tile {
    constructor() {
        this.solid = false;
        this.color = null;
    }
}
if (false) {
    /** @type {?} */
    Tile.prototype.solid;
    /** @type {?} */
    Tile.prototype.color;
}
export class GameManagerService {
    constructor() {
        this.settings = {
            tileSize: null
        };
        this._gridSize = { width: 0, height: 0 };
        this._locked = true;
        this._lineCleared = new Subject();
        this._gameOver = new Subject();
        this.lineCleared$ = this._lineCleared.asObservable();
        this.gameOver$ = this._gameOver.asObservable();
    }
    /**
     * @return {?}
     */
    get elementsInRow() {
        return this._gridSize.width;
    }
    /**
     * @param {?} width
     * @param {?} height
     * @param {?} gameSpeed
     * @param {?=} tileSize
     * @return {?}
     */
    initialize(width, height, gameSpeed, tileSize) {
        this._gridSize.width = width;
        this._gridSize.height = height;
        this._gameSpeed = gameSpeed;
        this._piecesFactory = new PiecesFactory(this._gridSize);
        if (tileSize) {
            this.settings.tileSize = tileSize;
        }
        this._initializeEmptyBoard();
        this._spawnNewPiece();
        this._drawPiece();
    }
    /**
     * @return {?}
     */
    start() {
        clearInterval(this._gameInterval);
        this._gameInterval = setInterval(() => {
            this._update();
        }, this._gameSpeed);
        this._locked = false;
    }
    /**
     * @return {?}
     */
    stop() {
        this._locked = true;
        clearInterval(this._gameInterval);
    }
    /**
     * @return {?}
     */
    reset() {
        /** @type {?} */
        const emptyTile = new Tile();
        for (let pos = 0; pos < this.grid.length; pos++) {
            if (this.grid[pos].color || this.grid[pos].solid) {
                this.__changeCell(pos, emptyTile);
            }
        }
        this._spawnNewPiece();
        this._drawPiece();
    }
    /**
     * @return {?}
     */
    moveLeft() {
        if (this._locked) {
            return;
        }
        this._clearPiece();
        this._piece.store();
        this._piece.moveLeft();
        if (this._collidesLeft()) {
            this._piece.revert();
        }
        this._drawPiece();
    }
    /**
     * @return {?}
     */
    moveRight() {
        if (this._locked) {
            return;
        }
        this._clearPiece();
        this._piece.store();
        this._piece.moveRight();
        if (this._collidesRight()) {
            this._piece.revert();
        }
        this._drawPiece();
    }
    /**
     * @return {?}
     */
    rotate() {
        if (this._locked) {
            return;
        }
        this._clearPiece();
        this._piece.store();
        this._piece.rotate();
        while (this._collidesRight()) {
            this._piece.moveLeft();
            if (this._collidesLeft()) {
                this._piece.revert();
                break;
            }
        }
        this._drawPiece();
    }
    /**
     * @return {?}
     */
    moveDown() {
        this._update();
    }
    /**
     * @return {?}
     */
    _clearFullLines() {
        for (let row = this._gridSize.height - 1; row >= 0; row--) {
            /** @type {?} */
            let isFull = true;
            for (let col = 0; col < this._gridSize.width; col++) {
                /** @type {?} */
                const pos = row * this._gridSize.width + col;
                if (this.grid[pos].solid === false) {
                    isFull = false;
                    break;
                }
            }
            if (isFull) {
                /** @type {?} */
                const emptyRow = Array.apply(null, Array(this._gridSize.width))
                    .map((idx) => new Tile());
                /** @type {?} */
                const topPortion = this.grid.slice(0, row * this._gridSize.width);
                this.grid.splice(0, ++row * this._gridSize.width, ...emptyRow.concat(topPortion));
                this._lineCleared.next();
            }
        }
    }
    /**
     * @return {?}
     */
    _update() {
        if (this._locked) {
            return;
        }
        this._locked = true;
        this._piece.revert();
        this._clearPiece();
        this._piece.store();
        this._piece.moveDown();
        if (this._collidesBottom()) {
            this._piece.revert();
            this._markSolid();
            this._drawPiece();
            this._clearFullLines();
            this._spawnNewPiece();
            if (this._isGameOver()) {
                this._onGameOver();
                return;
            }
        }
        this._drawPiece();
        this._locked = false;
    }
    /**
     * @return {?}
     */
    _isGameOver() {
        this._piece.store();
        this._piece.moveDown();
        if (this._collidesBottom()) {
            return true;
        }
        this._piece.revert();
        return false;
    }
    /**
     * @return {?}
     */
    _onGameOver() {
        this.stop();
        this._gameOver.next();
    }
    /**
     * @return {?}
     */
    _spawnNewPiece() {
        this._piece = this._piecesFactory.getRandomPiecePiece(SPAWN_POSITION_X, SPAWN_POSITION_Y);
    }
    /**
     * @return {?}
     */
    _initializeEmptyBoard() {
        /** @type {?} */
        const cellsCount = this._gridSize.width * this._gridSize.height;
        this.grid = Array.apply(null, Array(cellsCount))
            .map((idx) => new Tile());
    }
    /**
     * @return {?}
     */
    _clearPiece() {
        this._piece.positionsOnGrid
            .forEach((pos) => {
            this.__changeCell(pos, { color: undefined });
        });
    }
    /**
     * @return {?}
     */
    _drawPiece() {
        this._piece.clearStore();
        this._piece.positionsOnGrid
            .forEach((pos) => {
            this.__changeCell(pos, { color: this._piece.color });
        });
    }
    /**
     * @return {?}
     */
    _markSolid() {
        this._piece.positionsOnGrid.forEach((pos) => {
            this.__changeCell(pos, { solid: true });
        });
    }
    /**
     * @param {?} pos
     * @param {?=} data
     * @return {?}
     */
    __changeCell(pos, data = {}) {
        this.grid[pos] = Object.assign({}, this.grid[pos], data);
    }
    /**
     * @return {?}
     */
    _collidesBottom() {
        if (this._piece.bottomRow >= this._gridSize.height) {
            return true;
        }
        return this.__collides();
    }
    /**
     * @return {?}
     */
    _collidesLeft() {
        if (this._piece.leftCol < 0) {
            return true;
        }
        return this.__collides();
    }
    /**
     * @return {?}
     */
    _collidesRight() {
        if (this._piece.rightCol >= this._gridSize.width) {
            return true;
        }
        return this.__collides();
    }
    /**
     * @return {?}
     */
    __collides() {
        return this._piece.positionsOnGrid
            .some((pos) => {
            if (pos > 0 && this.grid[pos] && this.grid[pos].solid) {
                return true;
            }
            return false;
        });
    }
}
GameManagerService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
GameManagerService.ctorParameters = () => [];
if (false) {
    /** @type {?} */
    GameManagerService.prototype.settings;
    /** @type {?} */
    GameManagerService.prototype.grid;
    /** @type {?} */
    GameManagerService.prototype.lineCleared$;
    /** @type {?} */
    GameManagerService.prototype.gameOver$;
    /** @type {?} */
    GameManagerService.prototype._gridSize;
    /** @type {?} */
    GameManagerService.prototype._piece;
    /** @type {?} */
    GameManagerService.prototype._piecesFactory;
    /** @type {?} */
    GameManagerService.prototype._locked;
    /** @type {?} */
    GameManagerService.prototype._gameSpeed;
    /** @type {?} */
    GameManagerService.prototype._gameInterval;
    /** @type {?} */
    GameManagerService.prototype._lineCleared;
    /** @type {?} */
    GameManagerService.prototype._gameOver;
}
