// ngx-tetris/esm5/board/board.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, Input, ViewChild } from '@angular/core';
import { GameManagerService } from '../services/game-manager.service';
import { TileComponent } from "../tile/tile.component";
var BoardComponent = /** @class */ (function () {
    function BoardComponent(_manager) {
        this._manager = _manager;
        this.boardWidth = null;
    }
    /**
     * @return {?}
     */
    BoardComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        if (this._manager.settings.tileSize) {
            /** @type {?} */
            var bb = this.tile.el.nativeElement.getBoundingClientRect();
            /** @type {?} */
            var TILE_MARGIN = 1;
            /** @type {?} */
            var oneTileSpace = bb.width + 2 * TILE_MARGIN;
            this.boardWidth = oneTileSpace * this._manager.elementsInRow;
        }
    };
    BoardComponent.decorators = [
        { type: Component, args: [{
                    selector: 'board',
                    template: "<tile *ngFor=\"let tileData of grid\"\n    [data]=\"tileData\">\n</tile>\n",
                    host: {
                        '[style.width]': 'boardWidth + "px"',
                    },
                    styles: [":host{width:290px;display:block}"]
                }] }
    ];
    /** @nocollapse */
    BoardComponent.ctorParameters = function () { return [
        { type: GameManagerService }
    ]; };
    BoardComponent.propDecorators = {
        grid: [{ type: Input }],
        tile: [{ type: ViewChild, args: [TileComponent,] }]
    };
    return BoardComponent;
}());
export { BoardComponent };
if (false) {
    /** @type {?} */
    BoardComponent.prototype.grid;
    /** @type {?} */
    BoardComponent.prototype.tile;
    /** @type {?} */
    BoardComponent.prototype.boardWidth;
    /** @type {?} */
    BoardComponent.prototype._manager;
}
