// angular2-hotkeys/__ivy_ngcc__/index.js
"use strict";
var ɵngcc0 = require('@angular/core');
var ɵngcc1 = require('./src/hotkeys.directive');
var ɵngcc2 = require('./src/cheatsheet.component');
var ɵngcc3 = require('@angular/common');
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var common_1 = require("@angular/common");
var hotkeys_directive_1 = require("./src/hotkeys.directive");
var cheatsheet_component_1 = require("./src/cheatsheet.component");
var hotkey_options_1 = require("./src/hotkey.options");
var hotkeys_service_1 = require("./src/hotkeys.service");
__export(require("./src/cheatsheet.component"));
__export(require("./src/hotkey.model"));
__export(require("./src/hotkey.options"));
__export(require("./src/hotkeys.directive"));
__export(require("./src/hotkeys.service"));
var HotkeyModule = HotkeyModule_1 = (function () {
    function HotkeyModule() {
    }
    HotkeyModule.forRoot = function (options) {
        if (options === void 0) { options = {}; }
        return {
            ngModule: HotkeyModule_1,
            providers: [
                hotkeys_service_1.HotkeysService,
                { provide: hotkey_options_1.HotkeyOptions, useValue: options }
            ]
        };
    };
HotkeyModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: HotkeyModule });
HotkeyModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function HotkeyModule_Factory(t) { return new (t || HotkeyModule)(); }, imports: [[common_1.CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(HotkeyModule, { declarations: [ɵngcc1.HotkeysDirective, ɵngcc2.CheatSheetComponent], imports: [ɵngcc3.CommonModule], exports: [ɵngcc1.HotkeysDirective, ɵngcc2.CheatSheetComponent] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(HotkeyModule, [{
        type: core_1.NgModule,
        args: [{
                imports: [common_1.CommonModule],
                exports: [hotkeys_directive_1.HotkeysDirective, cheatsheet_component_1.CheatSheetComponent],
                declarations: [hotkeys_directive_1.HotkeysDirective, cheatsheet_component_1.CheatSheetComponent]
            }]
    }], function () { return []; }, null); })();
    return HotkeyModule;
}());
exports.HotkeyModule = HotkeyModule;
var HotkeyModule_1;
