// ngx-tetris/esm2015/public_api.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TetrisCoreComponent } from './tetris-core.component';
import { BoardComponent } from './board/board.component';
import { GameManagerService } from './services/game-manager.service';
import { TileComponent } from './tile/tile.component';
export { GameState, TetrisCoreComponent } from './tetris-core.component';
export class TetrisCoreModule {
    /**
     * @return {?}
     */
    static forRoot() {
        return {
            ngModule: TetrisCoreModule
        };
    }
}
TetrisCoreModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule
                ],
                declarations: [
                    TetrisCoreComponent,
                    BoardComponent,
                    TileComponent
                ],
                providers: [
                    GameManagerService
                ],
                exports: [TetrisCoreComponent]
            },] }
];
