// ngx-tetris/esm5/tile/tile.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, ElementRef, Input, Renderer2, ViewEncapsulation } from '@angular/core';
import { GameManagerService, Tile } from '../services/game-manager.service';
var TileComponent = /** @class */ (function () {
    function TileComponent(el, _renderer, _manager) {
        this.el = el;
        this._renderer = _renderer;
        this._manager = _manager;
        this.tileSize = null;
        if (this._manager.settings.tileSize) {
            this.tileSize = this._manager.settings.tileSize;
        }
    }
    /**
     * @return {?}
     */
    TileComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (this.data.color) {
            this._renderer.addClass(this.el.nativeElement, this.data.color);
        }
    };
    TileComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tile',
                    template: "",
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    // tslint:disable-next-line
                    host: {
                        '[style.width]': 'tileSize',
                        '[style.height]': 'tileSize'
                    },
                    styles: ["tile{display:block;background:rgba(176,230,225,.35);width:25px;height:25px;float:left;border:1px solid #4e4645;margin:1px}tile.color-box{background:#fca}tile.color-dot{background:#ff77a8}tile.color-line{background:#83769c}tile.color-t{background:#29adff}tile.color-l{background:#00e436}tile.color-lr{background:#ffec27}tile.color-z{background:#ffa300}tile.color-s{background:#c2c3c7}"]
                }] }
    ];
    /** @nocollapse */
    TileComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 },
        { type: GameManagerService }
    ]; };
    TileComponent.propDecorators = {
        data: [{ type: Input }]
    };
    return TileComponent;
}());
export { TileComponent };
if (false) {
    /** @type {?} */
    TileComponent.prototype.data;
    /** @type {?} */
    TileComponent.prototype.tileSize;
    /** @type {?} */
    TileComponent.prototype.el;
    /** @type {?} */
    TileComponent.prototype._renderer;
    /** @type {?} */
    TileComponent.prototype._manager;
}
